File: Problem230/Solution.java
public class Solution {
	private int kthSmallest;

	public int kthSmallest(TreeNode root, int k) {
		getKthElement(root, k, 0);
		return kthSmallest;
	}

	private int getKthElement(TreeNode root, int k, int curr) {
		if (root == null)
			return curr;
		curr = getKthElement(root.left, k, curr);
		curr++;
		if (curr == k) {
			kthSmallest = root.val;
			return curr;
		}
		return getKthElement(root.right, k, curr);
	}
}

File: Problem230/Solution.go
package main

type TreeNode struct {
	Val   int
	Left  TreeNode
	Right TreeNode
}

var smallestNum int

func kthSmallest(root *TreeNode, k int) int {
	getSmallest(root, k, 0)
	return smallestNum
}

func getSmallest(root *TreeNode, k int, curr int) int {
	if root == nil {
		return curr
	}
	curr = getSmallest(root.Left, k, curr)
	curr++
	if curr == k {
		smallestNum = root.Val
		return curr
	}

	return getSmallest(root.Right, k, curr)
}
